lib/accesslists/hierarchy.go

```diff
@@ -0,0 +1,569 @@
+	if exists {
+		// Reuse existing node's reverse pointers to parents.
+		tempNode = &HierarchyNode{
+			AccessList:    accessList,
+			MemberUsers:   make(map[string]struct{}),
+			MemberLists:   make(map[string]*HierarchyNode),
+			OwnerUsers:    make(map[string]struct{}),
+			OwnerLists:    make(map[string]*HierarchyNode),
+			MemberParents: existingNode.MemberParents,
+			OwnerParents:  existingNode.OwnerParents,
+		}
+	} else {
+		tempNode = &HierarchyNode{
+			AccessList:    accessList,
+			MemberUsers:   make(map[string]struct{}),
+			MemberLists:   make(map[string]*HierarchyNode),
+			OwnerUsers:    make(map[string]struct{}),
+			OwnerLists:    make(map[string]*HierarchyNode),
+			MemberParents: make(map[string]*HierarchyNode),
+			OwnerParents:  make(map[string]*HierarchyNode),
+		}
+	}
+
+	for _, owner := range accessList.Spec.Owners {
+		if owner.MembershipKind != accesslist.MembershipKindList {
+			tempNode.OwnerUsers[owner.Name] = struct{}{}
+			continue
+		}
+		ownerNode, ownerExists := h.Nodes[owner.Name]
+		if !ownerExists {
+			return trace.NotFound("Owner Access List '%s' not found", owner.Name)
+		}
+		if err := h.validateAddition(tempNode, ownerNode, RelationshipKindOwner); err != nil {
+			return trace.Wrap(err)
+		}
+		tempNode.OwnerLists[owner.Name] = ownerNode
+	}
+	for _, member := range members {
+		if member.Spec.MembershipKind != accesslist.MembershipKindList {
+			tempNode.MemberUsers[member.Spec.Name] = struct{}{}
+			continue
+		}
+		memberNode, memberExists := h.Nodes[member.Spec.Name]
+		if !memberExists {
+			return trace.NotFound("Member Access List '%s' not found", member.Spec.Name)
+		}
+		if err := h.validateAddition(tempNode, memberNode, RelationshipKindMember); err != nil {
+			return trace.Wrap(err)
+		}
+		tempNode.MemberLists[member.Spec.Name] = memberNode
+	}
+	return nil
+}
+
+// IsAccessListOwner determines if a User is a valid Owner of an existing or new AccessList,
+// including via inheritance. If User has any inForce Locks, it will return an error.
+func (h *Hierarchy) IsAccessListOwner(ctx context.Context, user types.User, accessListName string) (MembershipOrOwnershipType, error) {
+	// Allow for Locks to be nil when not provided in constructor.
+	if h.Locks != nil {
+		locks, err := h.Locks.GetLocks(ctx, true, types.LockTarget{
+			User: user.GetName(),
+		})
+		if err != nil {
+			return MembershipOrOwnershipTypeNone, trace.Wrap(err)
+		}
+
+		if len(locks) > 0 {
+			return MembershipOrOwnershipTypeNone, trace.AccessDenied("User '%s' is currently locked", user.GetName())
+		}
+	}
+
+	node, exists := h.Nodes[accessListName]
+	if !exists {
+		return MembershipOrOwnershipTypeNone, trace.NotFound("Access List '%s' not found", accessListName)
+	}
+
+	// Check explicit owners
+	if _, ok := node.OwnerUsers[user.GetName()]; ok {
+		// Verify ownership requirements using provided AccessList
+		if !userMeetsRequirements(user, node.AccessList.Spec.OwnershipRequires) {
```

Comment:
Where do we validate owner/member TTL? If an owner/member TTL expired, they are no longer considered valid owner/member. I don't see the TTL check anywhere here, unless I missed it.

Same question for members function below.

Reply:
Thanks for the catch, forgot to re-add that logic! Owners don't expire, but inherited Owners should be valid Members in their respective lists, so I added that logic in.